### PR TITLE
fix concurrent write on websocket

### PR DESCRIPTION
**Pull request by daniel-meyer-dme**
- fix locking in client
- fix locking in (websocket-) connection

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -190,8 +201,11 @@ func (ws *websocket) Read() (messageType int, msg []byte, err error) {
 	}
 
 	// ensure that we have the connection during the whole read operation
-	ws.mux.RLock()
-	defer ws.mux.RUnlock()
+	ws.read.Lock()
+
+	defer func() {
+		ws.read.Unlock()
```
Comment (ThomasObenaus) [inline connection.go:207]: why not just
`defer ws.read.Unlock()`

File: request.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ func prepareAuthRequest(requestID string, username string, password string) requ
 	req := request{}
 	req.RequestID = requestID
 	req.Op = "authentication"
-	req.Processor = "trasversal"
+	req.Processor = "traversal"
```
Comment (ThomasObenaus) [inline request.go:68]: how did this work beforehand? ... is the processor ignored for the auth operation?

File: request_test.go [hunk 3]
```diff
@@ -121,7 +121,7 @@ func TestAuthRequestPreparation(t *testing.T) {
 	testRequest := prepareAuthRequest(id, "test", "root")
 
 	assert.Equal(t, testRequest.RequestID, id)
-	assert.Equal(t, "trasversal", testRequest.Processor)
+	assert.Equal(t, "traversal", testRequest.Processor)
```
Comment (ThomasObenaus) [inline request_test.go:124]: :-) nice test

File: connection.go (hunk 1 above)
Comment (daniel-meyer-dme) [reply connection.go:207]: will remove, had some logging in there before

File: infra/gremlin-server.yaml [hunk 4]
```diff
@@ -30,13 +30,15 @@ scriptEngines: {
...
   - { className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0, config: { serializeResultToString: true }}                                                                       # application/vnd.gremlin-v3.0+gryo-stringd
-  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0, config: { ioRegistries: [org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}         # application/json
-  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphBinaryMessageSerializerV1 }                                                                                                            # application/vnd.graphbinary-v1.0
+  #- { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0, config: { ioRegistries: [org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}         # application/json
+  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV2d0, config: { ioRegistries: [ org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV2d0 ] } }         # application/json -> application/vnd.gremlin-v2.0+json
```
[the remark this answers is not in the document]
Comment (daniel-meyer-dme) [reply infra/gremlin-server.yaml:34]: see request.go l. 11-12
```
// MimeType used for communication with the gremlin server.
var MimeType = []byte("application/vnd.gremlin-v2.0+json")
```

File: infra/gremlin-server.yaml [hunk 5]
```diff
@@ -30,13 +30,15 @@ scriptEngines: {
 serializers:
   - { className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0, config: { ioRegistries: [org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}             # application/vnd.gremlin-v3.0+gryo
   - { className: org.apache.tinkerpop.gremlin.driver.ser.GryoMessageSerializerV3d0, config: { serializeResultToString: true }}                                                                       # application/vnd.gremlin-v3.0+gryo-stringd
-  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0, config: { ioRegistries: [org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}         # application/json
-  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphBinaryMessageSerializerV1 }                                                                                                            # application/vnd.graphbinary-v1.0
+  #- { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV3d0, config: { ioRegistries: [org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV3d0] }}         # application/json
```
[the remark this answers is not in the document]
Comment (daniel-meyer-dme) [reply infra/gremlin-server.yaml:33]: yes, you can only have one configuration for application/json and we specified to use v2 in request.go l. 11-12
```
// MimeType used for communication with the gremlin server.
var MimeType = []byte("application/vnd.gremlin-v2.0+json")
```

File: infra/gremlin-server.yaml [hunk 6]
```diff
@@ -30,13 +30,15 @@ scriptEngines: {
...
+  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphSONMessageSerializerV2d0, config: { ioRegistries: [ org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerIoRegistryV2d0 ] } }         # application/json -> application/vnd.gremlin-v2.0+json
+  - { className: org.apache.tinkerpop.gremlin.driver.ser.GraphBinaryMessageSerializerV1 }
+  # application/vnd.graphbinary-v1.0
 processors:
   - { className: org.apache.tinkerpop.gremlin.server.op.session.SessionOpProcessor, config: { sessionTimeout: 28800000 }}
   - { className: org.apache.tinkerpop.gremlin.server.op.traversal.TraversalOpProcessor, config: { cacheExpirationTime: 600000, cacheMaxSize: 1000 }}
 metrics: {
-  consoleReporter: {enabled: true, interval: 180000},
+  consoleReporter: {enabled: false, interval: 180000},
```
[the remark this answers is not in the document]
Comment (daniel-meyer-dme) [reply infra/gremlin-server.yaml:41]: noise in the logs, will add it again

File: request_test.go (hunk 3 above)
Comment (daniel-meyer-dme) [reply request_test.go:124]: indeed

File: connection.go [hunk 7]
```diff
@@ -27,6 +27,7 @@ type websocket struct {
 
 	// connected flags the websocket as connected or not connected
 	connected bool
+	connectedMux sync.Mutex
```
[the remark this answers is not in the document]
Comment (daniel-meyer-dme) [reply connection.go:30]: done

File: request.go (hunk 2 above)
Comment (daniel-meyer-dme) [reply request.go:68]: There is no test that uses auth, so I guess it is ignored.

**Comment by mattKorwel** [comment]
Hey Way to go all. I totally let you hanging on my issue and you got it knocked out anyway. thanks much!

**Pull request merged**